test/onnx/test_pytorch_onnx_onnxruntime.py

```diff
@@ -978,88 +963,124 @@ def forward(self, x, y=None, z=None):
-        self.run_test(MixedModel(), (x, y, None), input_names=["input_x", "input_y"])
-        self.run_test(MixedModel(), (x, None, z), input_names=["input_x", "input_z"])
-
-        # With optional arguments dictionary
-        self.run_test(MixedModel(), (x, {"y": y, "z": None}), input_names=["input_x", "input_y"])
-        self.run_test(MixedModel(), (x, {"y": None, "z": z}), input_names=["input_x", "input_z"])
+        model = Model()
+        # Without kwargs dict.
+        self.run_test(model, (x, y, None))
+        self.run_test(model, (x, None, z))
+
+        # With kwargs dict.
+        # Requires input_names to be set so that we can feed the inputs properly into ORT.
+
+        # TODO: Export default values as ONNX initializers, then this should not raise.
+        # https://msdata.visualstudio.com/Vienna/_workitems/edit/969268
+        # Default values are accessible via FunctionSchema.
+        with self.assertRaisesRegex(ValueError, "Model requires 3 inputs. Input Feed contains 2"):
+            self.run_test(model, (x, {"y": y}), input_names=("x", "y"))
+
+        # Setting any input to None in tracing means that
+        # input doesn't exist at all in the exported model, so input_names doesn't match.
+        # Thus the remaining tests only work in scripting.
+        # In tracing we'll provide the user with a reasonable error message:
+        # "number of input names provided (3) exceeded number of inputs (2)"
+        script_model = torch.jit.script(model)
+        self.run_test(script_model, (x, {"y": y, "z": None}), input_names=("x", "y", "z"))
+        self.run_test(script_model, (x, {"y": None, "z": z}), input_names=("x", "y", "z"))
 
-        class AllOptionalModel(torch.nn.Module):
-            def forward(self, y=None, z=None):
+    @skipScriptTest()  # Needs https://github.com/pytorch/rfcs/pull/21
+    @skipIfUnsupportedMinOpsetVersion(15)
+    def test_all_optional_default_none(self):
+        class Model(torch.nn.Module):
+            def forward(self, x: Optional[Tensor] = None, y: Optional[Tensor] = None):
+                if x is not None:
+                    return x
                 if y is not None:
                     return y
-                if z is not None:
-                    return z
+                else:
+                    return torch.tensor(-1.)
 
-        y = torch.randn(2, 3)
-        z = torch.randn(2, 3)
-        # Without optional arguments dictionary
-        self.run_test(AllOptionalModel(), (y, None), input_names=["input_y"])
-        self.run_test(AllOptionalModel(), (None, z), input_names=["input_z"])
-        # With optional arguments dictionary
-        self.run_test(AllOptionalModel(), {"y": y, "z": None}, input_names=["input_y"])
-        self.run_test(AllOptionalModel(), {"y": None, "z": z}, input_names=["input_z"])
-
-    def test_input_as_output(self):
+        x = torch.randn(2, 3)
+        model = Model()
+        self.run_test(model, (x, None))
```

Comment:
With tracing, do these tests capture `if` operator in exported graph?

I wonder if it'd work when providing extra set of inputs (e.g. (None, other_tensor)) to `run_test`.

Reply:
from comments looks like the answer to above question is it won't capture if and it won't work with other inputs combination in tracing ... 

Is it reasonable to assume that for optional input feature to really be effective, it would require scripting?

Reply:
Thanks for explanation. Let's keep it to ensure tracing route is compatible with this signature and None/Optional usage.